### PR TITLE
feat: Direct call w/o value to system accounts [0.0.361-0.0.750] change output

### PR DESCRIPTION
**Pull request by stefan-stefanooov**
Direct call w/o value to system accounts [0.0.361-0.0.750] result in precompile error

### **Description**:
Executing a ContractCall w/o value to system account in the range[0.0.361-0.0.750) result in PRECOMPILE_ERROR
`Expected result:` The call should `succeed` with `no-op`.

### **Steps to reproduce**

1. Create an account
2. Execute ContractCall to 0.0.629 using the account from step1.
3. `Expected` result: The call should succeed with no-op.
4. `Actual` result: The call fails with PRECOMPILE_ERROR

**Related issue(s)**:
https://github.com/hashgraph/hedera-services/issues/11033
Fixes #
https://github.com/hashgraph/hedera-services/issues/11033

Fixes #

### **Notes for reviewer**

This is achieved by adding a new function called `handleNonExtantSystemAccount`. It will: 

>  //    Create an empty/successful result object
> `PrecompileContractResult result = PrecompileContractResult.success(Bytes.EMPTY);`
> //    Pass it to the existing **finishPrecompileExecution** function that mutates the **frame** object with the required values.
> `finishPrecompileExecution(frame, result, PRECOMPILE, (ActionSidecarContentTracer) tracer);`

We call this new function in `CustomMessageCallProcessor.java` -> `start`, when: 

> - We don't have a Hedera system contract (precompile) with this input address **codeAddress** -> `systemContracts.containsKey(codeAddress)` returns `false`
> - the input **codeAddress** for the frame is from the reserved system accounts rage [0.0.0-0.0.750) -> `addressChecks.isSystemAccount` returns `true`
> - we don't have a holt event
> - we don't have a EVM precompile with this input address
>

### CODE REVIEW

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/processors/CustomMessageCallProcessor.java [hunk 1]
```diff
@@ -94,7 +93,7 @@ public CustomMessageCallProcessor(
     }
 
     /**
-     * Starts the execution of a message call based on the contract address of the given frame,
+     * Sta§rts the execution of a message call based on the contract address of the given frame,
```
Comment (lukelee-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/processors/CustomMessageCallProcessor.java:96]: Is this change on purpose?

**Review by lukelee-sl — COMMENTED** [review]
I wonder if this use case is handled by [this](https://github.com/hashgraph/hedera-services/pull/12842/files) pr?

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/processors/CustomMessageCallProcessor.java [hunk 2]
```diff
@@ -124,12 +123,23 @@ public void start(@NonNull final MessageFrame frame, @NonNull final OperationTra
             return;
         }
 
+        // Handle evm precompiles
```
Comment (lukelee-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/processors/CustomMessageCallProcessor.java:126]: Question - why did this block have to move.  If there is a good explanation maybe include a comment?

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/processors/CustomMessageCallProcessor.java (hunk 1 above)
Comment (stefan-stefanooov) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/processors/CustomMessageCallProcessor.java:96]: Not really. My mistake ... it should have been a "$" I guess... for all the money that is going to start flowing after this fix is deployed. 💰 
I'll remove it. Thank you, Sir.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/processors/CustomMessageCallProcessor.java (hunk 2 above)
Comment (stefan-stefanooov) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/processors/CustomMessageCallProcessor.java]: I moved it because of readability and changing operation order was important. 
But since then for consolidating the work with the MR you mentioned above ... I changed back the order. Thank you.

**Comment by stefan-stefanooov** [comment]
> I wonder if this use case is handled by [this](https://github.com/hashgraph/hedera-services/pull/12842/files) pr?

That was a good catch. 
We synced with the Merge Request authors and we are figuring out a way to move forward. 
We are gravitating towards  merging my Request first and then they will work on top of it. 

Thank you Sir!

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/junit/HapiTestEngine.java [hunk 3]
```diff
@@ -217,7 +217,8 @@ public HapiTestEngineExecutionContext before(HapiTestEngineExecutionContext cont
 
             // Allow for a simple switch to enable in-process Alice node for debugging
             final String debugEnv = System.getenv("HAPI_DEBUG_NODE");
-            final boolean debugMode = Boolean.parseBoolean(debugEnv);
+            // final boolean debugMode = Boolean.parseBoolean(debugEnv);
+            final boolean debugMode = true;
```
Comment (Neeharika-Sompalli) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/junit/HapiTestEngine.java:221]: Is this intentional ?

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/traceability/TraceabilitySuite.java [hunk 4]
```diff
@@ -4245,7 +4241,7 @@ HapiSpec traceabilityE2EScenario15() {
                                                             .setRecipientContract(
                                                                     spec.registry()
                                                                             .getContractId(contract))
-                                                            .setGasUsed(80135)
+                                                            .setGasUsed(80193)
```
Comment (Neeharika-Sompalli) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/traceability/TraceabilitySuite.java:4244]: Are these gas changes different from mono-service ?

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/traceability/TraceabilitySuite.java [hunk 5]
```diff
@@ -4260,7 +4256,7 @@ HapiSpec traceabilityE2EScenario15() {
                                                             .setCallingContract(
                                                                     spec.registry()
                                                                             .getContractId(contract))
-                                                            .setGas(3870609)
+                                                            .setGas(3870552)
```
Comment (Neeharika-Sompalli) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/traceability/TraceabilitySuite.java:4259]: same here

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/traceability/TraceabilitySuite.java [hunk 6]
```diff
@@ -4588,7 +4580,7 @@ final HapiSpec traceabilityE2EScenario21() {
                                                 .setCallingAccount(TxnUtils.asId(GENESIS, spec))
                                                 .setCallOperationType(CallOperationType.OP_CALL)
                                                 .setGas(978936)
-                                                .setGasUsed(963748)
+                                                .setGasUsed(3172)
```
Comment (Neeharika-Sompalli) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/traceability/TraceabilitySuite.java:4583]: Does this fail mono-service and work for mod service ? Why are we changing these numbers?

Comment (stefan-stefanooov) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/traceability/TraceabilitySuite.java:4583]: Hey! We are changing the flow and returning different statuses to the user. I tough that was the reason for the change in gas. Am I wrong ?

Comment (stefan-stefanooov) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/traceability/TraceabilitySuite.java:4583]: @Neeharika-Sompalli

Comment (Neeharika-Sompalli) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/traceability/TraceabilitySuite.java:4583]: I am not sure. But changing `HapiSpec` seems like changing the logic from mono-service. 
 @tinker-michaelj  thoughts?

**Review by petreze — APPROVED** [review]

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/traceability/TraceabilitySuite.java (hunk 6 above)
Comment (stefan-stefanooov) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/traceability/TraceabilitySuite.java:4590]: @lukelee-sl , @david-bakin-sl
Could you please review this ?

Comment (lukelee-sl) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/traceability/TraceabilitySuite.java:4590]: IMO, we should discourage (but still allow) this use case where the user tries to call a function on a system account.  One way to do that is if we charge all the remaining gas in the transaction but still allow the transaction to succeed.  I believe if we did this then the `gasUsed` value above in the test may not have to be updated (I haven't tested this).

Comment (david-bakin-sl) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/traceability/TraceabilitySuite.java:4590]: If we do this `system-accounts-operations.md` should be updated - it doesn't discuss fees for "no-op success" at all.

Comment (petreze) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/traceability/TraceabilitySuite.java:4590]: @Neeharika-Sompalli @tinker-michaelj @lukelee-sl @david-bakin-sl 
We are changing the logic from mono, indeed. This fix and others from the same package are bugs found in mono codebase and has to be fixed in modu

Comment (stefan-stefanooov) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/traceability/TraceabilitySuite.java]: New version available guys :) 
In it we consume all the gas. 
https://github.com/hashgraph/hedera-services/pull/12833/files#diff-79d792e4de7dd21f5fe7e60edebaea811a3bfd113a27838a84208e9747c00cbbR174

**Review by lukelee-sl — APPROVED** [review]
LGTM. Thank you!

**Review by petreze — APPROVED** [review]

**Review by Neeharika-Sompalli — APPROVED** [review]
I am not sure if the chnages to Hapi tests are expected. Will defer to @tinker-michaelj  or @lukelee-sl  or @david-bakin-sl review 🙂
